blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/UserSSHKeyManager.java

```
@@ -0,0 +1,214 @@
+        for(CredentialsStore s : CredentialsProvider.lookupStores(user)) {
+            if(s.hasPermission(CredentialsProvider.CREATE) && s.hasPermission(CredentialsProvider.UPDATE)){
+                return s;
+            }
+        }
+        return null;
+    }
+    
+    /**
+     * Gets an "ssh-rsa"-style formatted public key with useful identifier
+     */
+    private static String getSSHPublicKey(RSAPublicKey key, User user) throws IOException {
+        return "ssh-rsa " + Base64.encode(SSHKeyUtils.encodePublicKey(key)) + " " + getKeyComment(user.getId());
+    }
+
+    /**
+     * Returns a <user>@<jenkins-host> sort of comment to help identify the key's origin
+     * @param userId a user id
+     * @return an identifier
+     */
+    private static String getKeyComment(String userId) {
+        String host = Jenkins.getInstance().getRootUrl();
+        if (host == null) {
+            host = Jenkins.getInstance().getRootUrlFromRequest();
+        }
+        host = host.replaceAll(".*//([^/]+).*", "$1");
+        return ((userId == null ? Jenkins.getInstance().getDisplayName() : userId) + "@" + host)
+                .replaceAll("[^:@._a-zA-Z0-9]", "");
+    }
+    
+    private static Domain getDomain(CredentialsStore store) {
+        Domain domain = store.getDomainByName(BLUEOCEAN_DOMAIN_NAME);
+        if (domain == null) {
+            try {
+                //create new one
+                boolean result = store.addDomain(new Domain(BLUEOCEAN_DOMAIN_NAME, null, null));
+                if (!result) {
+                    throw new RuntimeException(String.format("Failed to create credential domain: %s", BLUEOCEAN_DOMAIN_NAME));
+                }
+                domain = store.getDomainByName(BLUEOCEAN_DOMAIN_NAME);
+                if (domain == null) {
+                    throw new RuntimeException(String.format("Domain %s created but not found", BLUEOCEAN_DOMAIN_NAME));
```

better to throw `UnexpectedErrorException` for proper json rendering.